### DISCUSSION TITLE
API: context queries

### DISCUSSION THREAD

**Discussion by ktsaou** [netdata/netdata#12112] [Ideas]
@stelfrag I have a question about the context queries of the API.

Let's take an example:

This is a query to get the `net.net` context of all network interfaces from the registry:

```sh
# curl 'https://registry.my-netdata.io/api/v1/data?context=net.net&points=10&before=0&after=-3600&options=jsonwrap|absolute|aligned&format=json&group=average'  
{
   "api": 1,
   "id": "net.net",
   "name": "net.net",
   "view_update_every": 360,
   "update_every": 1,
   "first_entry": 1644242696,
   "last_entry": 1644502692,
   "before": 1644502680,
   "after": 1644499081,
   "dimension_names": ["received", "sent", "received", "sent"],
   "dimension_ids": ["received", "sent", "received", "sent"],
   "chart_ids": ["net.vpn0", "net.vpn0", "net.eth0", "net.eth0"],
   "latest_values": [0, 0, 152.1515692, -101.602063],
   "view_latest_values": [2.3656864, 0, 177.8589566, 230.913013],
   "dimensions": 4,
   "points": 10,
   "format": "json",
   "result": {
 "labels": ["time", "received", "sent", "received", "sent"],
    "data":
 [
      [ 1644502680, 2.3656864, 0, 177.8589566, 230.913013],
      [ 1644502320, 3.0973238, 0, 178.8174752, 234.0979086],
      [ 1644501960, 2.420643, 0.0015556, 241.082902, 249.3213201],
      [ 1644501600, 1.7989107, 0, 181.404258, 255.9037317],
      [ 1644501240, 2.9663318, 0, 183.7845692, 243.5694234],
      [ 1644500880, 3.2396554, 0, 187.1236171, 264.6914973],
      [ 1644500520, 3.6822969, 0, 181.2079066, 239.4728118],
      [ 1644500160, 2.5123004, 0, 183.8695478, 241.8341782],
      [ 1644499800, 1.8818442, 0, 177.1267165, 233.9350358],
      [ 1644499440, 1.7134265, 0, 177.5188331, 244.8658513]
  ]
},
 "min": -264.6914973,
 "max": 241.082902
}
```

We see that it returned 5 dimensions:

1. the timestamp
2. `net.vpn0/received`
3. `net.vpn0/sent`
4. `net.eth0/received`
5. `net.eth0/sent`

Note that we requested `group=average`, so for every of the 10 points received, the agent aggregated `"view_update_every": 360` database points, with the `sum` function.

All good so far.

Now we run the same query, but we request back a flattened array:

```sh
# curl 'https://registry.my-netdata.io/api/v1/data?context=net.net&points=10&before=0&after=-3600&options=jsonwrap|absolute|aligned&format=array&group=average' 
{
   "api": 1,
   "id": "net.net",
   "name": "net.net",
   "view_update_every": 360,
   "update_every": 1,
   "first_entry": 1644242697,
   "last_entry": 1644502693,
   "before": 1644502680,
   "after": 1644499081,
   "dimension_names": ["received", "sent", "received", "sent"],
   "dimension_ids": ["received", "sent", "received", "sent"],
   "chart_ids": ["net.vpn0", "net.vpn0", "net.eth0", "net.eth0"],
   "latest_values": [0.3820606, 0, 125.6404329, -220.1619118],
   "view_latest_values": [2.3656864, 0, 177.8589566, 230.913013],
   "dimensions": 4,
   "points": 10,
   "format": "array",
   "result": [2.3656864,3.0973238,2.4221986,1.7989107,2.9663318,3.2396554,3.6822969,2.5123004,1.8818442,1.7134265],
 "min": 1.7134265,
 "max": 3.6822969
}
```

Boom! What is this? How the dimensions where flattened? Which algorithm was used?
I was expecting a sum of all dimensions into one.

Note the `before` and `after` of both responses. They match.

So, how (the first row in the first response):

```
2.3656864, 0, 177.8589566, 230.913013
```

leads to (the first number of the second response):

```
2.3656864
```

Further questions:

1. The context query, does it go through multiple hosts of a parent, or limits itself to the host it queried?
2. Normally, it would be beneficial the flattening of multiple dimensions to one to happen using a different algorithm. But there is no option to do that. Can we add it?

**Comment by stelfrag** [comment]
- The context query is per host (so it will limit itself to the localhost or the host specified in the /**host**/api/v1/data)
- The `format=array` was not supported in the context queries, so the result is wrong! We will fix it
   - We can also add an option to specify how to flatten the result (not just sum)

**Comment by ktsaou** [comment]
Great! Do you want me to turn this into an issue. It is a button. Let me try it...

**Comment by ktsaou** [comment]
Done. Opened an issue: https://github.com/netdata/netdata/issues/12113

**Reply by stelfrag** [reply]
The fix has been applied. 

Now for the second part of specifying a **different algorithm** for the flattening of multiple dimensions, what algorithms would make sense
to allow?

**Reply by ktsaou** [reply]
So, we have already an algorithm `group=sum|average|min|max|etc...` for aggregating metric values (of the same metric) over time. (`group` is a bad name, probably it should be `time-group` to indicate this is only applied over time).

There is already another algorithm, that by the time I was maintaining was not parameterized, it was always `sum`. This algorithm was used to aggregate the dimensions of an `RRDR` to create just one series of values. It was indirectly used, when the `format` parameter required flattening of the `RRDR`. Let's call this algorithm `dimension-group`. Although `sum` is a good default for most of the cases, in a few cases like `sensors.temperature` it is totally wrong (what is the point of summing the temperatures?). Probably `min`, `max` and `average` would be better for them.

Then, there is a third case, introduced with context queries. How to aggregate multiple charts to one series of values. I haven't seen the code, so I don't know if this is really a third case. If you build just one `RRDR` from multiple charts, we could use the `dimension-group` here too. If you are building multiple `RRDR`, one for each chart, we could introduce a `chart-group` to flatten them.

To give you the most complex example I can think of:

The user has 10 network interfaces. Each network interface has a bandwidth chart with context `net.net` and dimensions `received` and `sent`. So 10 charts, 20 dimensions total.

The user wants a chart with just one metric, representing the maximum bandwidth used by any of he 10 network interfaces over the last hour, given in 10 time-series points. Just a line. 10 points.

The query will be `before=0&after=-3600&points=10`

`options=absolute` will also be used, so all values are positive (`sent` is negative by default).

Then:

1. `group=average` (or better `time-group=average`) will average each of the 20 dimensions over time. So, every dimension of each chart will be averaged over time to have just 10 points each. Still 20 time-series. This is an `RRDR` in the code.
   
2. `dimension-group=sum` will flatten each chart to one value, by summing `received` and `sent` into just one time series. So, now we have 10 time-series (not 20), one per chart, of 10 points each, representing the total bandwidth of each network interface over the last hour.
   
3. `chart-group=max` will find the max bandwidth of all interfaces, so the result will be 1 time-series on 10 points, representing the maximum bandwidth used by any of the 10 network interfaces over the last hour.

**Reply by ktsaou** [reply]
Better naming:

* `group` the algorithm for aggregating values of a dimension over time
* `flatten` the algorithm for aggregating multiple dimensions into one time-series, for the whole chart
* `context-flatten` the algorithm for aggregating multiple charts into one time-series, for the whole context

**Reply by stelfrag** [reply]
The way this currently works is by building one RRDR so there is no list of charts that it can separately apply `flatten` and then `context-flatten` 

So it will currently produce an output that includes all dimensions from all matching charts, that is for 10 interfaces --> 10 pairs of (sent, received)

To be able to use `group`, `flatten` and `context-flatten` we need to change the way the queries are executed (per chart) and
- Produce the same output as we do now in case no `format=array` (flattening is not requested)
- Produce the output you described otherwise (flatten on chart and on context)

**Reply by ktsaou** [reply]
@stelfrag I think is simpler than that.

We support `group` already.
We support `flatten` already but it is not configurable, it is always `SUM`.

If we do just the above, `flatten` will be configurable, but `context-flatten` will always be the same with `flatten`.

This is still good up to here!

---

Just an idea, to avoid refactoring and support `context-flatten` too:

Add to that 1 RRDR you have, another small array of N numbers, where N is the number of dimensions.

Then for each chart encountered, give it a number, so that in the array you will have per dimension the chart number that contributed that dimension.

Example:

network interfaces, context `net.net`, dimensions `received` and `sent`.

let's assume 3 charts.
The newly added array would have `1, 1, 2, 2, 3, 3`.

So the first 2 dimensions come from the first chart, the next 2 from the second, the last 2 from the third.

Then you could `flatten` the 6 dimensions into another `RRDR`, with 3 dimension (as many as the charts).
Then you could `context-flatten` the new RRDR with the 3 dimensions, into one array of numbers the user wants.

Anyway, this is not required for the moment. I think supporting just a configurable `flatten` will do the trick we currently need.
